File: 프로그래머스/lv2/49994. 방문 길이/방문 길이.js
const solution = (dirs) => {
    const [x, y, BOUNDARY] = [0, 1, 5];
    const inArea = (command, pos) =>
        (command === 'L' && pos[x] > -BOUNDARY)
        || (command === 'R' && pos[x] < BOUNDARY)
        || (command === 'D' && pos[y] > -BOUNDARY)
        || (command === 'U' && pos[y] < BOUNDARY)

    return [...dirs].reduce((acc, command) => {
        if (!inArea(command, acc.pos)) return acc;

        const prev = [...acc.pos];
        (command === 'L') ? --acc.pos[x] : (command === 'R') ? ++acc.pos[x] : 0;
        (command === 'D') ? --acc.pos[y] : (command === 'U') ? ++acc.pos[y] : 0;

        if (!acc.history[[prev, acc.pos]] && !acc.history[[acc.pos, prev]]) {
            acc.history[[prev, acc.pos]] = acc.history[[acc.pos, prev]] = true;
            ++acc.move;
        }
        return acc;
    }, { pos: [0, 0], history: {}, move: 0 }).move;
}

File: 프로그래머스/lv2/49994. 방문 길이/방문 길이.cpp
#include <string>
#include <unordered_map>
#include <set>
using namespace std;

#define POS 5

bool move(int& x, int& y, int moveX, int moveY)
{
    if (x + moveX < -POS || x + moveX > POS 
        || y + moveY < -POS || y + moveY > POS)
        return false;

    x += moveX;
    y += moveY;
    return true;
}

int solution(string dirs) {
    int answer = 0, x = 0, y = 0;
    enum { U, D, R, L };
    int dirX[] = { 0, 0, 1, -1 },
        dirY[] = { 1, -1, 0, 0 };
    unordered_map<int, set<int>> m;

    for (size_t i = 0; i < dirs.length(); ++i)
    {
        int dirIdx = -1, prevIdx = -1;

        if (dirs[i] == 'U') dirIdx = U, prevIdx = D;
        else if (dirs[i] == 'D') dirIdx = D, prevIdx = U;
        else if (dirs[i] == 'R') dirIdx = R, prevIdx = L;
        else if (dirs[i] == 'L') dirIdx = L, prevIdx = R;

        const int prevID = x * 100 + y; // 자신만의 이전 인덱스 고유 ID
        if (move(x, y, dirX[dirIdx], dirY[dirIdx]))
        {
            const int newID = x * 100 + y; // 자신만의 새로운 인덱스 고유 ID
            if (m.find(prevID) != m.end()
                && m[prevID].find(dirIdx) != m[prevID].end())
                continue;

            m[prevID].insert(dirIdx); // prev -> new 방향
            m[newID].insert(prevIdx); // new -> prev 방향
            ++answer;
        }
    }
    return answer;
}